### PR TITLE
Zeitwerk Loader Implementation

### PR DESCRIPTION
**Pull request by gadimbaylisahil**
https://github.com/bensheldon/good_job/issues/87

### CODE REVIEW

**Comment by gadimbaylisahil** [comment]
Logger currently has issues. Not sure why yet, will investigate later. 

GoodJob.logger doesn't change, due to initializer not being triggered I suppose.

File: spec/lib/good_job/cli_spec.rb [hunk 1]
```diff
@@ -1,6 +1,5 @@
 # frozen_string_literal: true
 require 'rails_helper'
-require 'good_job/cli'
```
Comment (gadimbaylisahil) [inline spec/lib/good_job/cli_spec.rb:3]: Not needed

File: good_job.gemspec [hunk 2]
```diff
@@ -43,6 +43,7 @@ Gem::Specification.new do |spec|
   spec.add_dependency "pg", ">= 1.0.0"
   spec.add_dependency "railties", ">= 5.1.0"
   spec.add_dependency "thor", ">= 0.14.1"
+  spec.add_dependency "zeitwerk", ">= 2.2.2"
```
Comment (gadimbaylisahil) [inline good_job.gemspec:46]: Not sure about the version range to depend on

File: lib/good_job.rb [hunk 3]
```diff
@@ -1,19 +1,17 @@
...
-require 'good_job/lockable'
-require 'good_job/job'
-require 'good_job/scheduler'
-require 'good_job/multi_scheduler'
-require 'good_job/adapter'
-require 'good_job/performer'
-require 'good_job/current_execution'
-require 'good_job/notifier'
+require "active_job"
+require "active_job/queue_adapters"
```
Comment (gadimbaylisahil) [inline lib/good_job.rb:4]: https://github.com/fxn/zeitwerk#reopening-third-party-namespaces

File: lib/good_job.rb [hunk 4]
```diff
@@ -1,19 +1,17 @@
...
+require "good_job/railtie"
```
Comment (gadimbaylisahil) [inline lib/good_job.rb:5]: When not requiring the Railtie the initializers were not triggered in railtie.rb file due to loading orders. 

To verify this assumption I did two things: 

1. Use `loader.eager_load` after `loader.setup` which worked.
2. Enable eager loading in test.rb file of spec/test_app which worked

Thus moved the railtie to good_job top requires. 

I usually require railtie on top of my main gem entry class as well. Opinions appreciated

File: lib/good_job.rb [hunk 5]
```diff
@@ -1,19 +1,17 @@
...
 
-require 'active_job/queue_adapters/good_job_adapter'
+require "zeitwerk"
 
+loader = Zeitwerk::Loader.for_gem
+loader.inflector.inflect(
+  'cli' => "CLI"
+)
+loader.push_dir(File.join(__dir__, ["generators"]))
```
Comment (gadimbaylisahil) [inline lib/good_job.rb:13]: Generator is not a namespace thus adding it to main directories

**Comment by gadimbaylisahil** [comment]
> Logger currently has issues. Not sure why yet, will investigate later.
> 
> GoodJob.logger doesn't change, due to initializer not being triggered I suppose.

Found the issue. See PR code comments

File: lib/good_job.rb (hunk 4 above)
Comment (bensheldon) [reply lib/good_job.rb:5]: I ran across this before in Zeitwerk being troublesome and did some [Github searching](https://github.com/search?l=&q=zeitwerk+%22railtie%22+path%3Alib&type=Code).

I think this should happen at the bottom of the file _after_ Zeitwerk is set up and _after_ the root namespace is created. Here's an example: https://github.com/dukex/aggregated_record/blob/e8a597aaee94eb6e03ba0a26654a9c7f50a9c321/lib/aggregated_record.rb#L4-L13

File: good_job.gemspec (hunk 2 above)
Comment (gadimbaylisahil) [reply good_job.gemspec]: changed

File: lib/good_job.rb (hunk 4 above)
Comment (gadimbaylisahil) [reply lib/good_job.rb]: Changed

**Pull request merged**